vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java

```diff
@@ -0,0 +1,99 @@
+import com.google.cloud.vertexai.generativeai.ChatSession;
+import com.google.cloud.vertexai.generativeai.FunctionDeclarationMaker;
+import com.google.cloud.vertexai.generativeai.GenerativeModel;
+import com.google.cloud.vertexai.generativeai.ResponseHandler;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+
+public class AutomaticFunctionCalling {
+  public static void main(String[] args) throws IOException, NoSuchMethodException {
+    // TODO(developer): Replace these variables before running the sample.
+    String projectId = "your-google-cloud-project-id";
+    String location = "us-central1";
+    String modelName = "gemini-1.5-flash-001";
+
+    String promptText = "What's the weather like in Paris?";
+
+    whatsTheWeatherLike(projectId, location, modelName, promptText);
+  }
+
+  /** Callable function getCurrentWeather. */
+  public static String getCurrentWeather(String location) {
+    if (location.equals("Paris")) {
+      return "raining";
+    } else {
+      return "sunny";
+    }
+  }
+
+  // A request involving the interaction with an external tool
+  public static String whatsTheWeatherLike(
+      String projectId, String location, String modelName, String promptText)
+      throws IOException, NoSuchMethodException {
+    // Initialize client that will be used to send requests.
+    // This client only needs to be created once, and can be reused for multiple requests.
+    try (VertexAI vertexAI = new VertexAI(projectId, location)) {
+      // Get the callable method instance
+      Method function = AutomaticFunctionCalling.class.getMethod("getCurrentWeather", String.class);
+      // Use the fromFunc helper method to create a FunctionDeclaration
+      FunctionDeclaration functionDeclaration =
+          FunctionDeclarationMaker.fromFunc(
+              "Get the current weather in a given location", function, "location");
+      System.out.println("Function declaration:");
+      System.out.println(functionDeclaration);
+
+      // Add the function to a "tool"
+      Tool tool = Tool.newBuilder().addFunctionDeclarations(functionDeclaration).build();
+
+      // Instantiate an AutomaticFunctionCallingResponder and add the callable method
+      AutomaticFunctionCallingResponder responder = new AutomaticFunctionCallingResponder();
+      responder.addCallableFunction("getCurrentWeather", function, "location");
+
+      // Start a chat session from a model, with the use of the declared function.
+      GenerativeModel model =
+          new GenerativeModel(modelName, vertexAI).withTools(Arrays.asList(tool));
+      ChatSession chat = model.startChat();
+
+      System.out.println(String.format("Ask the question: %s", promptText));
+      // Send message with the responder, which auto-responds FunctionCalls and
+      // returns the final text result
+      GenerateContentResponse response =
+          chat.withAutomaticFunctionCallingResponder(responder).sendMessage(promptText);
```

Comment:
If the `withAutomaticFunctionCallingResponder(responder)` is not chained here, will the function be auto-inferred? What's the use of chaining the function here when it's already present as part of the tool?

Reply:
No, the function won't be auto-inferred if `withAutomaticFunctionCallingResponder(responder)` is not presented.

While adding the tool introduces the "Function Calling" feature, enabling the model to potentially respond with a `FunctionCall` instruction, users still need to manually execute the function, construct a `FunctionResponse`, and make another request.

Adding the responder activates "Automatic Function Calling" feature, automating these manual steps within the client library, saving users time and code. However, as not all functions are callable within the same Java runtime, this feature is disabled by default, users need to explicit chain it when making the request.

Reply:
Thanks for the clarification!